skiko/src/iosMain/kotlin/org/jetbrains/skiko/SkikoUIView.kt

```diff
@@ -117,57 +122,360 @@ class SkikoUIView : UIView, UIKeyInputProtocol {
+     * If the text-range object is nil, it indicates that there is no current selection.
+     * https://developer.apple.com/documentation/uikit/uitextinput/1614541-selectedtextrange
+     */
+    override fun selectedTextRange(): UITextRange? {
+        return skiaLayer?.skikoView?.input?.getSelectedTextRange()?.toUITextRange()
+    }
+
+    /**
+     * The range of currently marked text in a document.
+     * If there is no marked text, the value of the property is nil.
+     * Marked text is provisionally inserted text that requires user confirmation;
+     * it occurs in multistage text input.
+     * The current selection, which can be a caret or an extended range, always occurs within the marked text.
+     * https://developer.apple.com/documentation/uikit/uitextinput/1614489-markedtextrange
+     */
+    override fun markedTextRange(): UITextRange? {
+        return skiaLayer?.skikoView?.input?.markedTextRange()?.toUITextRange()
+    }
+
+    override fun setMarkedTextStyle(markedTextStyle: Map<Any?, *>?) {
+        // do nothing
+    }
+
+    override fun markedTextStyle(): Map<Any?, *>? {
+        return null
+    }
+
+    /**
+     * Inserts the provided text and marks it to indicate that it is part of an active input session.
+     * Setting marked text either replaces the existing marked text or,
+     * if none is present, inserts it in place of the current selection.
+     * https://developer.apple.com/documentation/uikit/uitextinput/1614465-setmarkedtext
+     * @param markedText The text to be marked.
+     * @param selectedRange A range within markedText that indicates the current selection.
+     * This range is always relative to markedText.
+     */
+    override fun setMarkedText(markedText: String?, selectedRange: CValue<NSRange>) {
+        val (locationRelative, lengthRelative) = selectedRange.useContents {
+            location.toInt() to length.toInt()
         }
-        skiaLayer?.skikoView?.onTouchEvent(events.toTypedArray())
+        val relativeTextRange = SkikoTextRange(locationRelative, locationRelative + lengthRelative)
+        skiaLayer?.skikoView?.input?.setMarkedText(markedText, relativeTextRange)
+    }
+
+    /**
+     * Unmarks the currently marked text.
+     * After this method is called, the value of markedTextRange is nil.
+     * https://developer.apple.com/documentation/uikit/uitextinput/1614512-unmarktext
+     */
+    override fun unmarkText() {
+        skiaLayer?.skikoView?.input?.unmarkText()
+    }
+
+    override fun beginningOfDocument(): UITextPosition {
+        return IntermediateTextPosition(0)
+    }
+
+    /**
+     * The text position for the end of a document.
+     * https://developer.apple.com/documentation/uikit/uitextinput/1614555-endofdocument
+     */
+    override fun endOfDocument(): UITextPosition {
+        return IntermediateTextPosition(skiaLayer?.skikoView?.input?.endOfDocument() ?: 0)
+    }
+
+    /**
+     * Attention! fromPosition and toPosition may be null
```

Comment:
Could the parameters type have `?`  - `UITextPosition?` ?

Reply:
Unfortunately, if I make the type nullable (UITextPosition?), then the compilation fails with **wrong override**.
May be we can you some annotations tricks to fix this compilation error...